frontend/lib/src/components/widgets/BaseWidget/SharedWidgetLabel.tsx

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useState,
+} from "react"
+
+import { WidgetLabel as BaseWidgetLabel, LabelProps } from "./WidgetLabel"
+
+interface LabelContextType {
+  labelProps: LabelProps | null
+  setLabelProps: React.Dispatch<React.SetStateAction<LabelProps | null>>
+}
+
+const WidgetLabelContext = createContext<LabelContextType | null>(null)
+
+export const WidgetLabelRenderer = ({
```

Comment:
**suggestion:** I spent some more time thinking about this and I think the `useState/useEffect` pair is unnecessary overhead and may seem non-obvious. Instead, we can expose a portal and have the JSX elements render directly into them:

```tsx
import React, {
  createContext,
  FC,
  PropsWithChildren,
  useContext,
  useRef,
} from "react"

import ReactDOM from "react-dom"

import { WidgetLabel as BaseWidgetLabel, LabelProps } from "./WidgetLabel"

interface LabelContextType {
  portalRef: React.RefObject<HTMLDivElement>
}

const WidgetLabelContext = createContext<LabelContextType | null>(null)

export const WidgetLabelRenderer: FC<PropsWithChildren> = ({ children }) => {
  const portalRef = useRef<HTMLDivElement>(null)

  return (
    <WidgetLabelContext.Provider value={{ portalRef }}>
      <div ref={portalRef} />
      {children}
    </WidgetLabelContext.Provider>
  )
}

export const WidgetLabel: FC<LabelProps> = labelProps => {
  const context = useContext(WidgetLabelContext)

  if (!context) {
    throw new Error("WidgetLabel must be used within a WidgetLabelRenderer")
  }

  const { portalRef } = context

  if (!portalRef.current) {
    throw new Error("WidgetLabelContext ref expected to exist!")
  }

  return ReactDOM.createPortal(
    <BaseWidgetLabel {...labelProps} />,
    portalRef.current
  )
}
```

Reply:
amazing!